### PR TITLE
fix: first cut at admin clients api

### PR DESCRIPTION
**Pull request by mikelax**
# Overview

This PR adds support for a few new admin api endpoints, specifically creating clients, and getting a client.

## Your PR Checklist 🚨

❤️ Please review the [guidelines for contributing](../docs/CONTRIBUTING.md) to this repository. Our goal is to merge PRs fast 💨 .

- [x] Check your code additions will fail neither code linting checks nor unit tests.
- [x] Additional units tests have been added to prove code updates work and fixes are effective
- [x] Additional documentation has been added (if appropriate)
- [x] Update Assignee field to yourself

## Issue Reference

Relates to #13 

## Open Questions or Items to Callout

1. One change that was introduced with this PR is to require a `tenant-name` to be supplied when first creating a new Client. This forces a Client to be associated with a tenant upon being created, instead of having this be a two-step operation. I felt this is better and will make the actual usage and management of The Usher data easier. 
2. I added requirements for the `client_id` attribute when creating a new Client to only contain lowercase alpha, numbers, dash and underscore. This makes it "url friendly", are we happy with this new requirement?

### CODE REVIEW

File: server/src/api_endpoints/clients/index.js
Comment (k1ch) [inline server/src/api_endpoints/clients/index.js:28]:
```suggestion
    if (!client?.client_id) {
      throw new Error(`Error creating client: ${client}`)
    }
```

**Comment by k1ch** [comment]
Regarding your open questions: 
1. Why do we need to associate a client with a tenant? Isn't it better to consider tenants and clients as two independent entities? 
  a. Tenant as identity provider of personas
  b. Clients, as entities that can be accessed through roles and permissions
2. Sounds like a good enhancement.

**Comment by mikelax** [comment]
@k1ch  I believe this PR is now ready for review. 
I added the DELETE client API along with additional tests. 
In a future PR we can handle any changes related to using key vs `client_id`.

File: database/layer/admin-client.js
Comment (k1ch) [inline database/layer/admin-client.js:28]: @mikelax This means you are requiring a client to always be associated with at least one tenant upon creation. Don't we want to keep these two entities independent of each other?

File: server/the-usher-openapi-spec.yaml
Comment (k1ch) [inline server/the-usher-openapi-spec.yaml:520]: Although we have 
```sql
CONSTRAINT clients_client_id_uq UNIQUE (client_id)
```
But isn't is better to rely on `clientKey` for the sake of consistency across the APIs?

File: database/layer/admin-client.js
Comment (mikelax) [reply database/layer/admin-client.js:28]: @k1ch  yes, my thinking is that there really isn't a use case where we would want a client to exist that isn't attached to at least one tenant, so might as well force this at time of create. I think this type of functionality is fine as this is the typical use case. Otherwise we are requiring the caller to make two APIs calls in succession. 

The similar functional question will be when we implement an API to remove a client from a tenant (not necessarily delete a Client) do we enforce a Client to be attached to at least one Tenant.

File: server/the-usher-openapi-spec.yaml
Comment (mikelax) [reply server/the-usher-openapi-spec.yaml:520]: @k1ch  this is the API that I was talking about in our conversation yesterday when discussing the use of client id vs. key. 

These APIs follow the existing convention that was from the start of the project of using client id as the "identifier". 
This is what I was asking if we should review and merge this PR (assuming everything else is ok barring the id vs. key) then as a separate change consider the question of client id vs. key.

Comment (k1ch) [reply server/the-usher-openapi-spec.yaml:520]: I am on board with the usage of `client_id` as the unique identifier. Additionally, I suggest we drop the `key` column, make `client_id` the primary key, and update the corresponding foreign keys. Having two unique identifiers in a table does not make sense to me

File: server/src/api_endpoints/clients/index.js
Comment (k1ch) [inline server/src/api_endpoints/clients/index.js:16]:
```suggestion
  const {
    tenant_name: tenantName,
    client_id: clientId,
    name,
    description,
    secret
  } = req.body
```

File: database/layer/admin-client.js
Comment (k1ch) [inline database/layer/admin-client.js:22]: Using an object instead of multiple arguments probably makes the function easier to use.

File: server/src/api_endpoints/clients/index.js
Comment (k1ch) [inline server/src/api_endpoints/clients/index.js:30]: I'm not a fan of using `clientname` and `clientdescription`. It seems unnecessarily verbose in naming
```suggestion
    res.status(201).send(client)
```

Comment (mikelax) [reply server/src/api_endpoints/clients/index.js:30]: @k1ch I updated the field `description`, as that is what it's supposed to be. 
We are currently using `clientname` in our datalake-api code so this would be a breaking change, unless we want to break consistency and have the new APIs return `name` and the existing `/self/clients` have a different response. 

Otherwise that would probably be a change by itself.

File: database/layer/admin-client.js
Comment (mikelax) [reply database/layer/admin-client.js:22]: Can we have code styling / formatting changes like this be a totally separate PR from new API functionality?

File: server/src/api_endpoints/clients/index.js
Comment (k1ch) [reply server/src/api_endpoints/clients/index.js:30]: Correct, when we update `/self/clients` to be consistent and use `name` instead of `clientname` that would be a breaking change. However, for this new API, I believe we can use `name` without any issue.

File: server/the-usher-openapi-spec.yaml
Comment (mikelax) [reply server/the-usher-openapi-spec.yaml:584]: @k1ch I see where you are going with the thought, but in this case I am not sure the refactor is worth the effort. 

Logically, having a "database unique key" and a different "Human / User unique key" is duplicative I don't think its the end of the world. If we were designing the schema from scratch we might do it differently.

File: database/layer/admin-client.js
Comment (k1ch) [reply database/layer/admin-client.js]: Yes, good idea.

File: server/test/endpoint_roles.test.js
Comment (k1ch) [inline server/test/endpoint_roles.test.js:22]: Are we planning to implement these tests under this PR? 
```suggestion
    it('Should create a new role')
    it('Should return a 409 for a duplicate role within a client')
```

File: server/test/endpoint_roles.test.js
Comment (k1ch) [inline server/test/endpoint_roles.test.js:23]: Is it bad request because `tenant_name` is missing? why is the `client_id` invalid? I believe `foo` passes the regex check `^[a-z0-9-_]*$`.

File: server/the-usher-openapi-spec.yaml
Comment (k1ch) [inline server/the-usher-openapi-spec.yaml:607]: Does it delete the associated permissions too?

Comment (mikelax) [reply server/the-usher-openapi-spec.yaml:607]: @k1ch  I just looked at the db schema and the `clientkey` column in the permissions table is not actually a foreign key to clients, therefore there is no on cascade setting currently. This should probably be a separate enhancement to make.

File: server/test/endpoint_roles.test.js
Comment (mikelax) [reply server/test/endpoint_roles.test.js:23]: @k1ch  this is a bad request because the client id `foo` does not exist. This test is trying to create a new Role, so it needs a valid client id to associate the Role with.

File: server/test/endpoint_roles.test.js
Comment (mikelax) [reply server/test/endpoint_roles.test.js:22]: @k1ch  I will implement these few tests. I have to leave the error code as 400 until we do a further refactor the error handling for these db functions to make use of your new knex patterns.

File: server/src/api_endpoints/clients/index.js
Comment (mikelax) [reply server/src/api_endpoints/clients/index.js]: @k1ch ok made the change, will commit shortly. The `/self/clients` actually only returns the client id and name so I think its fair to consider that one API a one-off.

File: server/test/endpoint_clients.test.js
Comment (k1ch) [inline server/test/endpoint_clients.test.js:12]: If we are not using function context `this`, we can use arrow function instead. We can be consistent in the file. 
```suggestion
  before(async () => {
```

File: server/test/endpoint_roles.test.js
Comment (k1ch) [inline server/test/endpoint_roles.test.js:63]: I see that we are not deleting both roles, as the next test is dependent on it. Shouldn't we ensure that each test is fully independent to avoid a chain failure if one test fails or to prevent deadwoods in the database?

File: server/test/endpoint_roles.test.js
Comment (k1ch) [inline server/test/endpoint_roles.test.js:88]: We should use `409` for conflict
```suggestion
      assert.strictEqual(res2.status, 409)
```

**Review by k1ch — DISMISSED** [review]
Very nice! Left some comments for improvement.

File: server/test/endpoint_roles.test.js
Comment (mikelax) [reply server/test/endpoint_roles.test.js:63]: @k1ch  the "first" role is being deleted by the `afterEach` function down around line 100. I know it's a bit weird to manually delete one item in the test, and then rely on an `afterEach` to delete the other. Maybe I will just add a comment to make this clear.

File: server/test/endpoint_roles.test.js
Comment (mikelax) [reply server/test/endpoint_roles.test.js:88]: @k1ch  I think that we need to make this change as a separate PR. In order to handle this error code we should refactor the db functions to return errors the new way and then it should be easier to get more accurate error codes for this API.

**Review by k1ch — APPROVED** [review]

**Pull request merged**
